lib/src/command/add.dart

```diff
@@ -0,0 +1,339 @@
+
+  bool get hasGitOptions => gitUrl != null || gitRef != null || gitPath != null;
+  bool get hasHostOptions => hostUrl != null;
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds package to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package in repository');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+    argParser.addOption('sdk', help: 'SDK source for package');
+
+    argParser.addFlag('offline',
+        help: 'Use cached packages instead of accessing the network.');
+
+    argParser.addFlag('dry-run',
+        abbr: 'n',
+        negatable: false,
+        help: "Report what dependencies would change but don't change any.");
+
+    argParser.addFlag('precompile',
+        help: 'Precompile executables in immediate dependencies.');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final packageInformation = _parsePackage(argResults.rest.first);
+    final package = packageInformation.first;
+
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackageToPubspec(entrypoint.root.pubspec, package);
+    final solveResult = await resolveVersions(
+        SolveType.GET, cache, Package.inMemory(updatedPubSpec));
+
+    final resultPackage = solveResult.packages
+        .firstWhere((packageId) => packageId.name == package.name);
+
+    /// Assert that [resultPackage] is within the original user's expectations.
+    if (package.constraint != null &&
+        !package.constraint.allows(resultPackage.version)) {
+      dataError('${package.name} resolved to ${resultPackage.version} which '
+          'does not match the input ${package.constraint}! Exiting.');
+    }
+
+    if (isDryRun) {
+      /// Even if it is a dry run, run `acquireDependencies` so that the user
+      /// gets a report on the other packages that might change version due
+      /// to this new dependency.
+      final newRoot = Package.inMemory(updatedPubSpec);
+
+      await Entrypoint.global(newRoot, entrypoint.lockFile, cache)
+          .acquireDependencies(SolveType.GET,
+              dryRun: true, precompile: argResults['precompile']);
+    } else {
+      /// Update the pubspec. It is necessary to do this before calling
+      /// [acquireDependencies] (as opposed to calling [acquireDependencies] on
+      /// the in-memory pubspec and writing it later) because there will be
+      /// conflicts in other parts of [pub] if the pubspec is newer than
+      /// `pubspec.lock`.
+      _updatePubspec(resultPackage, packageInformation, isDev);
```

Comment:
Will there be conflicts?

Even if not, we could a resolution in memory, and write `pubspec.yaml` and then touch timestamp on the other files.

Or we could resolve then write all 3 files in the right order.

Reply:
ah maybe conflicts is not the right word... was thinking about causing `pub run` or other similar commands to have to resolve `pubspec.lock` again. Either way, because we run `acquireDependencies` after updating the `pubspec.yaml`, the timestamps on the other files should resolve correctly. :D

The purpose of the comment was more to explain why we don't do something like `acquireDependencies` before writing to the pubspec file!

Reply:
Okay, can we update the comment to say: 
> Update the `pubspec.yaml` before calling acquireDependencies to ensure that
> the modification timestamp on `pubspec.lock` and `.dart_tool/package_config.json` is
> newer than the `pubspec.yaml`, ensuring that [entrypoint.assertUptoDate] will pass.